runtime/compiler/optimizer/J9ValuePropagation.cpp

```diff
@@ -1153,40 +1154,40 @@ J9::ValuePropagation::constrainRecognizedMethod(TR::Node *node)
 
             flagsForTransform.set(ValueTypesHelperCallTransform::InsertDebugCounter);
 
-            if (isStoreFlattenableArrayElement && !owningMethodDoesNotContainStoreChecks(this, node))
+            if (isStoreFlattenableArrayElement)
                {
-               // Determine whether the value is being copied from the same array that is the target
```

Comment:
I think null value store check should be independent of array store check

Reply:
I think the idea was that `owningMethodDoesNotContainStoreChecks` is true for methods that are known to be working with arrays where copying values from one array to another or storing values into an array will never require run-time checks - likely because they are only used with arrays whose component type is `java/lang/Object` or perhaps because they are always known to be working with arrays whose components types are the same.

We should probably double-check, but I believe that the methods that won't trigger an `ArrayStoreCheckException` will also not trigger a `NullPointerException` if the value that's being stored is a null reference.  Maybe `owningMethodDoesNotContainStoreChecks` could be renamed to indicate that it's safe for both array checks and null pointer exceptions, or a second method could be introduced to indicate whether the method is known not to trigger null pointer exceptions for the value on `aastore` operations.

Reply:
>but I believe that the methods that won't trigger an ArrayStoreCheckException will also not trigger a NullPointerException if the value that's being stored is a null reference.


`owningMethodDoesNotContainStoreChecks` looks at if a method is in the table of [canSkipArrayStoreChecks](https://github.com/eclipse-openj9/openj9/blob/3e340dbee3b2b7d25162373e3bf88f5b41437f65/runtime/compiler/il/J9MethodSymbol.cpp#L380-L433). If it is, it is safe to skip array store check. 

I picked `java_util_Hashtable_put` from `canSkipArrayStoreChecks` as an example.  `java/util/Hashtable/put` does not need an array store check, but it could throw `NullPointerException`. This seems contradictory to the above observation. I guess `owningMethodDoesNotContainStoreChecks` cannot be used indicate it's safe for both array checks and null pointer exceptions. Is my understanding correct? @hzongaro  

```
     * @throws     NullPointerException  if the key or value is
     *               {@code null}
     * @see     Object#equals(Object)
     * @see     #get(Object)
     */
    public synchronized V put(K key, V value) {
...
```

Reply:
You're right that `Hashtable.put` might throw a `NullPointerException` if either the key or value is a null reference, but it does that with an explicit test in the case of the value[1] and an unprotected dereference in the case of the key.[2]

I think the intent of the `canSkipArrayStoreChecks` table was to tell the JIT that the implementations of these methods are known to be safe with respect to array store operations - they will never attempt to store a value into an array that is not permitted to be stored into the array, but it might be impossible (or very unlikely) for the JIT to determine that given the way the implementation is structured.

In skipping the check that a null reference might be stored into an array with a null-restricted component type for those same methods, I made an assumption - that the nature of the implementations of those same methods is such that either the arrays that are involved never have null-restricted component types or that they will never attempt to store a null reference into an array that has a null-restricted component type.

However, I don't believe that I ever carefully examined all of the methods that are involved to verify that - just a subset, which is bad.  I will take a to-do to perform that task.  Either way, I think the right thing would be to split `canSkipArrayStoreChecks` into two tables:  `canSkipArrayStoreChecks` and `canSkipNonNullableArrayNullValueChecks`, or something like that.  After I complete my investigation, it might be that both tables contain all the same methods, but it would allow for the possibility that the two tables could diverge in the future.

Thanks for raising this, Annabelle @a7ehuo!

[1] https://github.com/ibmruntimes/openj9-openjdk-jdk17/blob/95863e9b987805363b31b6049d70603cfe67ae01/src/java.base/share/classes/java/util/Hashtable.java#L475-L477
[2] https://github.com/ibmruntimes/openj9-openjdk-jdk17/blob/95863e9b987805363b31b6049d70603cfe67ae01/src/java.base/share/classes/java/util/Hashtable.java#LL481C6-L481C6

Reply:
Yes, it would be OK if it stays as you've modified it.  It could have an impact on performance, but it would be correct, whereas the original code might not be.  I have opened issue #17491 to track work on creating a new table.